src/components/sidenav/sidenav.ts

```diff
@@ -0,0 +1,193 @@
+import {
+  AfterContentInit,
+  Component,
+  ContentChildren,
+  ElementRef,
+  EventEmitter,
+  Input,
+  View,
+  ViewEncapsulation,
+  OnChanges,
+  Output,
+  Query,
+  QueryList,
+  SimpleChange
+} from 'angular2/core';
+import {BaseException} from 'angular2/src/facade/exceptions';
+import {isPresent} from 'angular2/src/facade/lang';
+
+
+// TODO(hansl): move this function to a helper.
+function getTransitionEnd_(element: any): string {
+  try {
+    const transitions: {[name: string]: string} = {
+      'WebkitTransition': 'webkitTransitionEnd',
+      'MozTransition': 'transitionend',
+      'OTransition': 'oTransitionEnd otransitionend',
+      'transition': 'transitionend'
+    };
+
+    for (const name of Object.keys(transitions)) {
+      if (element.style[name] !== undefined) {
+        return transitions[name];
+      }
+    }
+  } catch (e) {
+    return null;
+  }
+}
+
+export class MdMissingSidenavException extends BaseException {}
+export class MdDuplicatedSidenavException extends BaseException {
+  constructor(align: MdSidenavAlignEnum) {
+    super(`A sidenav was already declared for 'align=${MdSidenavAlignEnum[align]}'`);
+  }
+}
+
+
+export enum MdSidenavAlignEnum {
+  START = 0,
+  END,
+};
+
+
+@Component({
+  selector: 'md-sidenav-container > md-sidenav, *[md-sidenav-container] > md-sidenav',
+  templateUrl: './components/sidenav/drawer.html',
+  styleUrls: ['./components/sidenav/drawer.css'],
+  host: {
+    '[class.md-sidenav-opened]': 'opened',
+  },
+})
+export class MdSidenav implements OnChanges {
+  @Input() align: MdSidenavAlignEnum = MdSidenavAlignEnum.START;
+  @Input('opened') private opened_: boolean;
+
+  @Output('open') onOpen = new EventEmitter<void>();
+  @Output('close') onClose = new EventEmitter<void>();
+
+  private promise_: Promise<void>;
+  private isOpening_: boolean = false;
+  private isClosing_: boolean = false;
+
+  private onTransitionEnd_(elementRef: ElementRef): void {
```

Comment:
I'd generally like to push for each method have at least a brief description of what it's doing, e.g.,
```ts
/** Emit the appropriate open/close event when an animation has completed. */
private onTransitionEnd_() {
  ...
}
```